shell/platform/darwin/macos/framework/Source/FLEViewController.mm

```diff
@@ -16,6 +16,52 @@
 
 static const int kDefaultWindowFramebuffer = 0;
 
+namespace {
+
+/**
+ * State tracking for mouse events, to adapt between the events coming from the system and the
+ * events that the embedding API expects.
+ */
+struct MouseState {
+  /**
+   * Whether or not a kAdd event has been sent (or sent again since the last kRemove if tracking is
+   * enabled). Used to determine whether to send a kAdd event before sending an incoming mouse
+   * event, since Flutter expects pointers to be added before events are sent for them.
+   */
+  bool flutter_state_is_added = false;
+
+  /**
+   * Whether or not a kDown has been sent since the last kAdd/kUp.
+   */
+  bool flutter_state_is_down = false;
+
+  /**
+   * Wether or not mouseExited: was received while a button was down. Cocoa's behavior when dragging
+   * out of a tracked area is to send an exit, then keep sending drag events until the last button
+   * is released. If it was released inside the view, mouseEntered: is sent the next time the mouse
+   * moves. Flutter doesn't expect to receive events after a kRemove, so the kRemove for the exit
+   * needs to be delayed until after the last mouse button is released.
+   */
```

Comment:
Let's see if my understanding is correct:
- If a mouse is pressed in the tracking area, then dragged out (and keeps moving), Cocoa will keep sending drag events with out-of-area positions.
- If the mouse is then dragged back in without releasing buttons, it will send a mouseEntered while keeping sending drag events.

If the statements above are correct,
- Should we stop sending drag events with out-of-area positions to Flutter, since they should be considered illegal data? Also perhaps to accommodate shells that can not detect out-of-area moves. 
- Should we set `_mouseState.has_pending_exit` to `false` in `mouseEntered:`, so that the "drag back in" action is considered a normal drag and does not trigger `Reset`?

Reply:
> If a mouse is pressed in the tracking area, then dragged out (and keeps moving), Cocoa will keep sending drag events with out-of-area positions.

Yes.

> If the mouse is then dragged back in without releasing buttons, it will send a mouseEntered while keeping sending drag events.

No, there's no mouseEntered; it's weirdly asymmetrical. (In that scenario, you don't the mouseEntered until the next time you move the mouse after releasing the button.)

> Should we stop sending drag events with out-of-area positions to Flutter

Let's revisit this later, since the patch isn't changing that behavior and we'll need to carefully think about what exactly the right behavior is in the context of yet-to-be-designed functionality like drag-and-drop from Flutter to other apps. (It's not as simple as just not sending events, because we don't want to drop the pointer if Flutter knows a drag is happening in that scenario, but we also don't want, say, an in-window drop target at the edge of the window to stay highlighted.)

> Should we set _mouseState.has_pending_exit to false in mouseEntered:

No (per answer 2 above)